pkg/kubelet/kubelet.go

```diff
@@ -869,47 +870,53 @@ func (kl *Kubelet) makeEnvironmentVariables(pod *api.Pod, container *api.Contain
 		return result, err
 	}
 
-	for _, value := range container.Env {
+	// Determine the final values of variables in steps:
+	//
+	// 1.  Determine the final value of each variable which is not sourced by expanding `$(var)`
+	//     references to other variables in the .Value field; the sources of variables are the
+	//     declared variables of the container and the service environment variables
+	// 2.  If a source is defined for an environment variable, resolve the source
+	// 3.  Create the container's environment in the order variables are declared
+	// 4.  Add remaining service environment vars
+
+	tmpEnv := make(map[string]string)
+	mappingFunc := expansion.MappingFuncFor(tmpEnv, serviceEnv)
+	for _, envVar := range container.Env {
 		// Accesses apiserver+Pods.
 		// So, the master may set service env vars, or kubelet may.  In case both are doing
 		// it, we delete the key from the kubelet-generated ones so we don't have duplicate
 		// env vars.
 		// TODO: remove this net line once all platforms use apiserver+Pods.
-		delete(serviceEnv, value.Name)
-
-		runtimeValue, err := kl.runtimeEnvVarValue(value, pod)
-		if err != nil {
-			return result, err
+		delete(serviceEnv, envVar.Name)
+
+		runtimeVal := envVar.Value
+		if runtimeVal != "" {
+			// Step 1: expand variable references
+			runtimeVal = expansion.Expand(runtimeVal, mappingFunc)
+		} else if envVar.ValueFrom != nil && envVar.ValueFrom.FieldRef != nil {
+			// Step 2: resolve alternate env var sources
+			runtimeVal, err = kl.podFieldSelectorRuntimeValue(envVar.ValueFrom.FieldRef, pod)
+			if err != nil {
+				return result, err
+			}
 		}
 
-		result = append(result, kubecontainer.EnvVar{Name: value.Name, Value: runtimeValue})
+		tmpEnv[envVar.Name] = runtimeVal
```

Comment:
Is there any particular reason not to append this to the result right here, where it used to be? For readability? Because of concern about duplicated names?